### PR TITLE
first version

### PR DESCRIPTION
**Pull request by kcaipq**
simple code please review.

### CODE REVIEW

File: src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.concurrent.{CancellationException, Future => JavaFuture}
+
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.Try
+
+
+object FutureUtils {
+
+  implicit class FutureOps[T](jFuture: JavaFuture[T]) {
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala]: Its probably easier / better to handle Amazon async commands by passing in an  `AsyncHandler` that completes a promise

File: src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+      * Recursive polling may be bad, should use timed polling, or jus use the blocking one below
+      */
+    def asScala: Future[T] = {
+      @tailrec
+      def checkCompletion(): Any = {
+        if (jFuture.isCancelled) {
+          promise.failure(new CancellationException(s"Java Future is cancelled. $jFuture"))
+        } else if (jFuture.isDone) {
+          promise.complete(Try(jFuture.get))
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala]: like this is surely going to use 100% CPU until the future completes

Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala]: yes, that's why I left a comment to use timed polling. Your AsyncHandler sounds better.

File: src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala (hunk 1 above)
Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/FutureUtils.scala]: 👍

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  // async aws client uses a thread pool that reuses a fixed number of threads
+  // operating off a shared unbounded queue.
+  val client: AmazonCloudWatchAsync = AmazonCloudWatchAsyncClientBuilder
+    .standard()
+    .withRegion(region)
+    .withExecutorFactory(
+      new ExecutorFactory {
+        // don't use the default thread pool which configures 50 number of threads
+        override def newExecutor(): ExecutorService = Executors.newFixedThreadPool(2)
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala]: How often does this get called? 2 threads is pretty low

File: build.sbt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+scalaVersion := "2.11.8"
+
+lazy val root = (project in file(".")).settings(
+  fork in Test := true
+)
+
+val akkaVersion = "2.4.18"
+
+libraryDependencies += "com.typesafe.akka" % "akka-actor_2.11" % akkaVersion
```
Comment (tomverran) [inline build.sbt]: can use `%%` here rather than _2.11

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    case msg => log.warning(s"Unsupported message $msg received in MetricsLogger")
+  }
+
+  /**
+    * Produce the datums.
+    * Code is take from:
+    * https://github.com/philwill-nap/Kamon/blob/master/kamon-cloudwatch/
+    * src/main/scala/kamon/cloudwatch/CloudWatchMetricsSender.scala
+    */
+  private def datums(tick: TickMetricSnapshot): List[List[MetricDatum]] = {
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala]: I believe the plural of `datum` is the more familiar `data` 😄

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    for {
+      (groupIdentity, groupSnapshot) <- tick.metrics
+      groupDimension = new Dimension().withName(groupIdentity.category).withValue(groupIdentity.name)
+      groupDimensions = List(groupDimension).asJava
+      (metricIdentity, metricSnapshot) <- groupSnapshot.metrics
+      datum = new MetricDatum().withDimensions(groupDimensions).withMetricName(metricIdentity.name).withTimestamp(new Date())
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala]: I'd probably move this into the `yield`

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala (hunk 5 above)
Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala]: thanks for the education 👍

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    ).build()
+
+
+  override def receive: Receive = LoggingReceive {
+    case ShipMetrics(metrics) =>
+      val f = client.putMetricDataAsync(
+        new PutMetricDataRequest().withNamespace(nameSpace).withMetricData(metrics.asJava)
+      )
+      f.asScala.pipeTo(self)
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala]: so what will happen to pipeTo if it gets a failed future? Won't it just end up logging "unsupported message"?

Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala]: yes, we are not interested in the failure of it, hence all being forced into unsupported.

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala (hunk 5 above)
Comment (tomverran) [reply src/main/scala/com/timeout/kamon/cloudwatch/MetricsLogger.scala]: Also I'd probably flatten the `List[List[` before returning

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala (hunk 7 above)
Comment (tomverran) [reply src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala]: I think maybe we'd be interested in at least logging that it failed, if we wonder where all the metrics have gone for example

File: src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala (hunk 3 above)
Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/MetricsShipper.scala]: I am going to make it adjustable by the client who is using this lib.

File: src/main/resources/reference.conf [hunk 8]
```diff
@@ -0,0 +1,47 @@
+kamon {
+  cloudwatch {
+
+    # make sure AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are set correctly in environment variables
+
+    # namespace is the AWS Metrics custom name-space
+    name-space = kamon-cloudwatch
+    # AWS region
+    region = eu-west-1
```
Comment (tomverran) [inline src/main/resources/reference.conf]: this is probably also a job for the getCurrentRegion API call, so we'd only use this setting if not on EC2

File: build.sbt (hunk 4 above)
Comment (ostapneko) [reply build.sbt]: we should probably cross-compile

File: src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import akka.event.Logging
+import com.timeout.kamon.cloudwatch.KamonSettings.subscriptions
+import com.typesafe.config.ConfigList
+
+import scala.collection.JavaConverters._
+import kamon.Kamon
+
+
+class KamonCloudWatchExtension(system: ExtendedActorSystem) extends Kamon.Extension {
+  val log = Logging(system, classOf[KamonCloudWatchExtension])
```
Comment (ostapneko) [inline src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala]: should be protected

File: src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  log.info("Starting the Kamon CloudWatch extension")
+
+  implicit val ec = system.dispatcher
+
+  val shipper: ActorRef = system.actorOf(MetricsShipper.props, "cloudwatch-metrics-shipper")
+  val subscriber: ActorRef = system.actorOf(MetricsLogger.props(shipper), "cloudwatch-metrics-logger")
+
+  subscriptions.entrySet.asScala.foreach { entry =>
```
Comment (ostapneko) [inline src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala]: Nitpick: it feels like most of the Java to Scala gymnastic should be done in `KamonSettings` instead

File: src/main/scala/com/timeout/kamon/cloudwatch/AmazonAsync.scala [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+object AmazonAsync {
+
+  type MetricDatumBatch = List[MetricDatum]
+
+  implicit class MetricsAsyncOps(data: MetricDatumBatch) {
+
+    def put(implicit client: AmazonCloudWatchAsync): Future[AmazonWebServiceResult[ResponseMetadata]] =
+      asyncRequest(new PutMetricDataRequest().withNamespace(nameSpace).withMetricData(data.asJava))(client.putMetricDataAsync)
+
+    private def asyncRequest[Arg, Req <: AmazonWebServiceRequest, Res](asyncArg: Arg)
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/AmazonAsync.scala]: ah you probably don't need this quite tricky method if you're only doing the one request, you can just make it non generic which'll make it more readable (i.e. just create a promise & an AsyncHandler for PutMetricData

File: src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+class KamonCloudWatchExtension(system: ExtendedActorSystem) extends Kamon.Extension {
+  implicit val ec: ExecutionContextExecutor = system.dispatcher
+  protected val log = Logging(system, classOf[KamonCloudWatchExtension])
+  log.info("Starting the Kamon CloudWatch extension")
+
+  val shipper: ActorRef = system.actorOf(MetricsShipper.props, "cloudwatch-metrics-shipper")
+  val subscriber: ActorRef = system.actorOf(MetricsLogger.props(shipper), "cloudwatch-metrics-logger")
+
+  // loop through patterns to register
+  subscriptions.foreach { entry => entry._2.foreach { pattern =>
```
Comment (tomverran) [inline src/main/scala/com/timeout/kamon/cloudwatch/KamonCloudWatchExtension.scala]: `case (name, patterns) =>` might aid readability here

File: src/main/scala/com/timeout/kamon/cloudwatch/AmazonAsync.scala (hunk 11 above)
Comment (kcaipq) [reply src/main/scala/com/timeout/kamon/cloudwatch/AmazonAsync.scala]: yes, this should be public and belong to AmazonAsync object which can be used by various types of async aws commands, i.e. MetricAsyncOps, S3AsyncOps, etc

**Comment by kcaipq** [comment]
if no more review I am going to merge it. @ostapneko @tomverran 
Published to local and tested with DQT. It worked fine.

**Comment by tomverran** [comment]
go for it 👍

**Pull request merged**